packages/client/src/reactions/Reactions.ts

```diff
@@ -1,37 +1,47 @@
 import { GraphQLClient } from 'graphql-request';
 
-import { Credentials } from '../authentication';
+import { Authentication } from '../authentication';
 import { LensConfig } from '../consts/config';
 import { NotAuthenticatedError } from '../consts/errors';
 import { ReactionRequest } from '../graphql/types.generated';
 import { getSdk, Sdk } from './graphql/reactions.generated';
 
 export class Reactions {
+  private readonly authentication: Authentication | undefined;
   private readonly sdk: Sdk;
 
-  constructor(config: LensConfig, private readonly credentials?: Credentials) {
+  constructor(config: LensConfig, authentication?: Authentication) {
     const client = new GraphQLClient(config.environment.gqlEndpoint);
 
-    if (credentials) {
-      client.setHeader('authorization', `Bearer ${credentials.accessToken}`);
-    }
-
     this.sdk = getSdk(client);
+    this.authentication = authentication;
   }
 
   async add(request: ReactionRequest) {
-    if (!this.credentials) {
+    if (!this.authentication) {
       throw new NotAuthenticatedError();
     }
 
-    return this.sdk.AddReaction(request);
+    const headerResult = await this.authentication.getRequestHeader();
+
+    if (headerResult.isFailure()) {
+      throw headerResult.unwrap();
```

Comment:
In case the `Result<T, E>` is ` Failure<T, E>`, `unwrap already throws the inner error, so technically you are just doing all of this with line 31.

I think we should have a cohesive narrative around known failure modes throughout the SDK. As we are exposing `Result<T, E>` in the "write hooks", I would strongly entertain the idea of doing the same here. It will allow for a more explicit error handling which can help with SDK upgrades (imagine a new error being returned or removed). Consumer's code if written in TS could be automatically informed about the change by simple TS and I would say very self-descriptive errors.

Reply:
I changed it to return `PromiseResult` but it's quite a lot of code, especially to change one definition of `PromiseResult` into another (with catching errors while unwrapping to wrap them again). Is there any better way to do it?

```
async add(
  request: ReactionRequest,
): PromiseResult<void, CredentialsExpiredError | NotAuthenticatedError | Error> {
  if (!this.authentication) {
    return failure(new NotAuthenticatedError());
  }

  const headerResult = await this.authentication.getRequestHeader();

  if (headerResult.isFailure()) {
    try {
      headerResult.unwrap();
    } catch (error) {
      assertError(error);
      return failure(error);
    }
  }

  try {
    await this.sdk.AddReaction(request, headerResult.unwrap());
  } catch (error) {
    assertError(error);
    return failure(error);
  }

  return success();
}
```

Reply:
I see, yeah you don't need to unwrap it at all if you are to use the result.

```ts
async add(
  request: ReactionRequest,
): PromiseResult<void, CredentialsExpiredError | NotAuthenticatedError | Error> {
  if (!this.authentication) {
    return failure(new NotAuthenticatedError());
  }

  const headerResult = await this.authentication.getRequestHeader();

  if (headerResult.isFailure()) {
    return failure(error);
  }

  await this.sdk.AddReaction(request, headerResult.value); // better not unwrap so if the `if` above gets mistakenly removed you get a TS error cause the result was not narrowed down to a Success<T, E> and `.value` becomes illegal.
  return success();
}
```

Reply:
I suspect after some iterations a code re-use pattern will emerge.

Reply:
check inline comments
```ts
// in your example above
if (headerResult.isFailure()) {
  return failure(error); // where error is coming from? 
}

```

```ts
try {
  // API call can also throw (ie. on connect issue), then the whole ResultPromise pattern doesn't make sense if it can still throw
  await this.sdk.AddReaction(request, headerResult.unwrap());
} catch (error) {
  assertError(error);
  return failure(error);
}
```

```

Reply:
Sorry the typo, the error is from the result:
```ts
if (headerResult.isFailure()) {
  return failure(headerResult.error);
}
```
**Side note:** as we do this relatively often an TS requires a new `failure` to reconcile the return type in case of `T` (in `Result<T, E>` mismatch).. we can consider a `forward` method in `Result`. It's something I saw in one `Result` implementation (cannot recall which one) but it was very neat.

The return line would change into:
```ts
return headingResult.forward()
```

Food for thoughts.

Reply:
done as discussed, thx!